sentinel/src/schedule/replications.js

```diff
@@ -37,11 +43,102 @@ function getSchedule(config) {
   }
 }
 
+const purgeDocs = (sourceDb, docs) => {
+  if (!docs || !docs.length) {
+    return Promise.resolve();
+  }
+
+  const docsToPurge = {};
+  docs.forEach(doc => docsToPurge[doc._id] = [doc._rev]);
+
+  return sourceDb
+    .info()
+    .then(info => {
+      const opts = {
+        uri: `${db.serverUrl}/${info.db_name}/_purge`,
+        json: true,
+        body: docsToPurge,
+      };
+
+      return rpn.post(opts);
+    });
+};
+
+const getHasRunOneTimePurge = (sourceDb) => {
+  return sourceDb
+    .get(ONE_TIME_PURGE_LOCAL_DOC_ID)
+    .then(() => true)
+    .catch(err => {
+      if (err.status === 404) {
+        return false;
+      }
+      throw err;
+    });
+};
+
+const oneTimePurgeRan = (sourceDb) => {
+  return sourceDb.put({ _id: ONE_TIME_PURGE_LOCAL_DOC_ID });
+};
+
+const oneTimePurge = (sourceDb, targetDb) => {
+  return getHasRunOneTimePurge(sourceDb)
+    .then(hasRunOneTimePurge => {
+      if (hasRunOneTimePurge) {
+        return;
+      }
+
+      return batchedPurge(sourceDb, targetDb).then(() => oneTimePurgeRan(sourceDb));
+    });
+};
+
+const batchedPurge = (sourceDb, targetDb, lastSeq = 0) => {
+  const opts = {
+    since: lastSeq,
+    limit: BATCH_SIZE,
+    batch_size: BATCH_SIZE,
+  };
+
+  const getReplicatedIds = (targetDb, ids) => {
+    if (!ids || !ids.length) {
+      return Promise.resolve([]);
+    }
+
+    return targetDb.changes({ doc_ids: ids }).then(result => result.results.map(change => change.id));
+  };
+
+  return sourceDb.changes(opts).then(result => {
+    if (!result.results.length) {
+      return;
+    }
+
+    const idsToPurge = result.results
+      .filter(change => !change.deleted && isTelemetryOrFeedback(change.id))
+      .map(change => change.id);
+
+    return getReplicatedIds(targetDb, idsToPurge).then(idsSafeToPurge => {
+      const docsToPurge = result.results
+        .filter(change => idsSafeToPurge.includes(change.id))
+        .map(change => ({ _id: change.id, _rev: change.changes[0].rev }));
+
+      return purgeDocs(sourceDb, docsToPurge).then(() => batchedPurge(sourceDb, targetDb, result.last_seq));
+    });
+  });
+};
+
+const isTelemetryOrFeedback = (docId) => docId.startsWith('telemetry-') || docId.startsWith('feedback-');
+
 function replicateDb(sourceDb, targetDb) {
   // Replicate only telemetry and feedback docs
-  return  sourceDb.replicate.to(targetDb, {
-    filter: doc => doc._id.startsWith('telemetry-') || doc._id.startsWith('feedback-')
-  });
+  return sourceDb.replicate
+    .to(targetDb, {
+      filter: doc => !doc._deleted && isTelemetryOrFeedback(doc._id),
+    })
+    .on('change', changes => {
+      return purgeDocs(sourceDb, changes.docs);
```

Comment:
Only running this can become a problem if a user logs in on multiple devices: 
Example: 
1. User logs in on device A and generates feedback doc X.
2. device A syncs doc X upwards. 
3. user logs in on device B and generates feedback doc Y. 
4. device B syncs and uploads doc Y and downloads doc X. 
5. the scheduler runs the replication and purges both doc X and doc Y. 
6. device B syncs again and uploads doc X. 

At this point, doc X will never be purged, because it's already replicated to the conglomerate (_revs_diff will yield no missing revs, because the rev is the same, so it will be skipped in replication). 

Unsure of the correct approach. 
Should we just purge all existent and replicated feedback docs every time (like the oneTime but everyTime)
and just store a seq of the last "thing" we purged instead of starting from 0 when reading the changes feed?

Reply:
I think it's OK. It's unlikely to happen often and so there will only be a little bloat but not real impact.